File: MatrizParallel/src/kotlinmat/Mult.kt
package kotlinmat

import kotlin.system.measureNanoTime

fun multiplyMatrices(a: Array<DoubleArray>, b: Array<DoubleArray>): Array<DoubleArray> {
    val result = Array(a.size) { DoubleArray(b[0].size) }
    for (i in a.indices) {              // Iterar sobre as linhas da matriz A
        for (j in b[0].indices) {       // Iterar sobre as colunas da matriz B
            result[i][j] = b.indices.sumOf { a[i][it] * b[it][j] }
        }
    }
    return result
}

fun main() {
    // Leitura das matrizes de arquivos CSV
    val a = readMatrixFromCsv("src/matrizes/matriz3.csv")
    val b = readMatrixFromCsv("src/matrizes/matriz4.csv")

    // Configurações iniciais
    val repetitions = 1             // Número de repetições para medir o tempo médio
    var totalNanoseconds = 0L       // Tempo total em nanosegundos

    // Imprimir o tamanho das matrizes
    println("Tamanho da matriz A: ${a.size}x${a[0].size}")
    println("Tamanho da matriz B: ${b.size}x${b[0].size}")

    // Repetir o processo de multiplicação de matrizes e medir o tempo médio de execução
    repeat(repetitions) {
        val elapsedTime = measureNanoTime {
            val finalResult = multiplyMatrices(a, b) // Multiplicação das matrizes sem paralelismo
        }
        totalNanoseconds += elapsedTime  // Acumula o tempo total
    }

    // Calcular a média
    val averageNanoseconds = totalNanoseconds / repetitions           // Tempo médio em nanosegundos
    val averageMilliseconds = averageNanoseconds / 1_000_000.0       // Tempo médio em milissegundos
    val averageSeconds = averageNanoseconds / 1_000_000_000.0        // Tempo médio em segundos

    // Exibir os resultados
    println("\nMédia de tempo de execução em segundos: %.10f s".format(averageSeconds))
    println("Média de tempo de execução em nanosegundos: $averageNanoseconds ns")
    println("Média de tempo de execução em milissegundos: $averageMilliseconds ms")

    // Imprimir o resultado final da multiplicação
    val finalResult = multiplyMatrices(a, b)
    println("\nResultado da multiplicação das matrizes:")
    //printMatrix(finalResult)
}

File: MatrizParallel/src/pythonmat/Mult.py
import time
import numpy as np

# Nesse exemplo, a multiplicação de matrizes é feita sem paralelismo!

def load_matrix_from_csv(filename):
    try:
        matrix = np.loadtxt(filename, delimiter=',')
        return matrix
    except Exception as e:
        print(f"Erro ao carregar o arquivo {filename}: {e}")
        return None

def multiply_matrices(a, b):
    result = []
    for i in range(len(a)):
        result_row = []
        for j in range(len(b[0])):
            result_row.append(sum(a[i][k] * b[k][j] for k in range(len(b))))
        result.append(result_row)
    return result

def measure_execution_time():
    # Carrega as matrizes dos arquivos CSV
    a = load_matrix_from_csv('src/matrizes/matriz3.csv')
    b = load_matrix_from_csv('src/matrizes/matriz4.csv')

    # Verificar se as matrizes foram carregadas corretamente
    if a is not None and b is not None:
        print(f"Tamanho da matriz a: {a.shape}")
        print(f"Tamanho da matriz b: {b.shape}")
    else:
        print("Erro ao carregar as matrizes. Verifique os arquivos CSV.")
        return

    num_runs = 1
    total_time = 0

    for _ in range(num_runs):
        start_time = time.perf_counter()

        # Multiplica as matrizes sem paralelismo
        final_result = multiply_matrices(a, b)

        end_time = time.perf_counter()
        total_time += (end_time - start_time)

    # Média de tempo de execução
    average_seconds      = total_time / num_runs
    average_milliseconds = average_seconds * 1000
    average_nanoseconds  = average_seconds * 1_000_000_000

    print(f"\nMédia de tempo de execução em segundos: {average_seconds:.10f} s")
    print(f"Média de tempo de execução em milissegundos: {average_milliseconds} ms")
    print(f"Média de tempo de execução em nanossegundos: {average_nanoseconds} ns")

    # Formatação dos resultados
    print("\nPrimeira linha do resultado da multiplicação:")
    #print(final_result[0])  # Exibe a primeira linha do resultado da multiplicação

if __name__ == "__main__":
    measure_execution_time()